manifest/dynamic_manifest.go

```diff
@@ -0,0 +1,226 @@
+
+	// the lk guards all dynamic manifest-specific fields
+	lk               sync.RWMutex
+	manifest         Manifest
+	onManifestChange OnManifestChange
+	nextManifest     *Manifest
+}
+
+func NewDynamicManifestProvider(manifest Manifest, pubsub *pubsub.PubSub, ec ec.Backend, manifestServerID peer.ID) ManifestProvider {
+	return &DynamicManifestProvider{
+		manifest:         manifest,
+		pubsub:           pubsub,
+		ec:               ec,
+		manifestServerID: manifestServerID,
+	}
+}
+
+func (m *DynamicManifestProvider) Manifest() Manifest {
+	m.lk.RLock()
+	defer m.lk.RUnlock()
+	return m.manifest
+}
+
+func (m *DynamicManifestProvider) GpbftOptions() []gpbft.Option {
+	return m.manifest.GpbftOptions()
+}
+
+func (m *DynamicManifestProvider) SetManifestChangeCallback(mc OnManifestChange) {
+	m.onManifestChange = mc
+}
+
+// When a manifest configuration changes, a new network name
+// is set that depends on the manifest version of the previous version to avoid
+// overlapping previous configurations.
+func (m *DynamicManifestProvider) networkNameOnChange() gpbft.NetworkName {
+	return gpbft.NetworkName(string(m.manifest.NetworkName) + "/" + fmt.Sprintf("%d", m.manifest.Sequence))
+}
+
+func (m *DynamicManifestProvider) Run(ctx context.Context, errCh chan error) {
+	if m.onManifestChange == nil {
+		errCh <- xerrors.New("onManifestChange is nil. Callback for manifest change required")
+	}
+	go m.handleIncomingManifests(ctx, errCh)
+	m.handleApplyManifest(ctx, errCh)
+}
+
+func (m *DynamicManifestProvider) handleApplyManifest(ctx context.Context, errCh chan error) {
+	// add a timer for EC period
+	ticker := time.NewTicker(ManifestCheckTick)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			m.lk.Lock()
+			if m.nextManifest != nil {
+				ts, err := m.ec.GetHead(ctx)
+				if err != nil {
+					log.Errorf("error fetching chain head: %+v", err)
+					m.lk.Unlock()
+					continue
+				}
+
+				// if the upgrade epoch is reached or already passed.
+				if ts.Epoch() >= m.nextManifest.BootstrapEpoch {
+					log.Debugf("reached bootstrap epoch, triggering manifest change: %d", ts.Epoch())
+					// update the current manifest
+					prevManifest := m.manifest
+					m.manifest = *m.nextManifest
+					nn := m.networkNameOnChange()
+					m.manifest.NetworkName = nn
+					m.nextManifest = nil
+					// trigger manifest change callback.
+					go m.onManifestChange(ctx, prevManifest, errCh)
+					m.lk.Unlock()
+					continue
+				}
+			}
+			m.lk.Unlock()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// listen to manifests being broadcast through the network.
+func (m *DynamicManifestProvider) handleIncomingManifests(ctx context.Context, errCh chan error) {
+	if err := m.setupManifestPubsub(); err != nil {
+		errCh <- xerrors.Errorf("setting up pubsub: %w", err)
+		return
+	}
+
+	manifestSub, err := m.manifestTopic.Subscribe()
+	if err != nil {
+		errCh <- xerrors.Errorf("subscribing to topic: %w", err)
+		return
+	}
+
+loop:
+	for {
```

Comment:
It would make the logic simpler to add a condition here to check if contex has error or not: `for ctx.Err() == nil {`